text/0000-layout-packed-aligned.md

```diff
@@ -0,0 +1,182 @@
+- Feature Name: `layout_packed_aligned`
+- Start Date: 2024-10-24
+- RFC PR: [rust-lang/rfcs#3718](https://github.com/rust-lang/rfcs/pull/3718)
+- Rust Issue: [rust-lang/rust#100743](https://github.com/rust-lang/rust/issues/100743)
+
+# Summary
+[summary]: #summary
+
+This RFC makes it legal to have `#[repr(C)]` structs that are:
+- Both packed and aligned.
+- Packed, and transitively contains`#[repr(align)]` types.
+
+It also introduces `#[repr(system)]` which is designed for interoperability with operating system APIs.
```

Comment:
No, the only difference between `#[repr(C)]` and `#[repr(system)]` by this RFC is that on pc-windows-gnu `#[repr(system)]` is the MSVC layout while `#[repr(C)]` is the GCC layout. On all other targets `#[repr(system)]` and `#[repr(C)]` are identical (per this RFC).

Reply:
I think `repr(linear)` is independently useful and we should have all 3. that said, I am concerned that current `repr(C)` code often means "I want stable linear layout" rather than "I want whatever weirdness the C compiler decides to use", so I think deprecating `repr(C)` and replacing it with `repr(linear)`, `repr(bikeshed_other_C)` and `repr(system)` is worth considering.

Reply:
Indeed, this RFC proposes a *different* distinction between `repr(C)` and `repr(system)` than what has been previously discussed in other threads.

Since the distinction between the two layouts we have here is Windows-only, I wonder if it should be some Windows-only name, like `repr(msvc)` or so? Is there a good reason to even make both of them available on all targets -- effectively exporting a Windows-only complication to other, saner platforms?

Reply:
> Is there a good reason to even make both of them available on all targets

The reason is that this is also how the `extern "system"` function ABI string works. Code uses `extern "system"` to link to libraries which use `__stdcall` on Windows platforms, and in the same way code would use `#[repr(system)]` for linking to a dylib which provides builds with only the MSVC toolchain for Windows targets.

I don't necessarily endorse this option, but it is logically consistent with how Rust already uses `"system"` as an ABI string.

Reply:
I do think that having `repr(linear)` or some kind of stable ABI valid for all platforms will be a good idea, although it's out-of-scope for this RFC. Additionally, this RFC will not impact those people already using `#[repr(C)]` for linear layouts, since it does not modify the behavior when you do not have `aligned` types nested within `packed` types, which was previously forbidden.

Reply:
It was previously forbidden but the check has [some gaps](https://github.com/rust-lang/rust/issues/100743#issuecomment-1229343705), so this *can* impact existing code.

It also breaks existing code that assumes that all `repr(C)` types are laid out according to the rules described [here](https://doc.rust-lang.org/reference/type-layout.html#the-c-representation).

Reply:
> The reason is that this is also how the extern "system" function ABI string works. Code uses extern "system" to link to libraries which use __stdcall on Windows platforms, and in the same way code would use #[repr(system)] for linking to a dylib which provides builds with only the MSVC toolchain for Windows targets.

From what I understand, `extern "system"` is the same on MSVC and GNU Windows targets? So this is IMO a false analogy then, making it more confusing than if we instead use a name that more explicitly represents that Windows has two ABIs, which we support with two target triples, and you might want to write code that talks with the "other" ABI.

Speaking of which, how would a program for the MSVC target lay out its type in the right way to call a GNU ABI library? That does not seem possible with this proposal.

Reply:
Yes, `extern "system"` is `extern "stdcall"` on all Windows targets. Just like how this RFC makes `repr(C)` on both windows-msvc and windows-gnu behave as `repr(MS)`. The difference is `repr(C)`, which switches between msvc/gnu respectively.

Reply:
Oops, yes, that's correct.